subsys/net/ip/net_if.c

```diff
@@ -1623,6 +1623,95 @@ bool net_if_ipv4_addr_rm(struct net_if *iface, struct in_addr *addr)
 
 	return false;
 }
+
+struct net_if_mcast_addr *net_if_ipv4_maddr_add(struct net_if *iface,
+						const struct in_addr *addr)
+{
+	int i;
+
+	if (!net_is_ipv4_addr_mcast(addr)) {
+		NET_DBG("Address %s is not a multicast address.",
+			net_sprint_ipv4_addr(addr));
+		return NULL;
+	}
+
+	for (i = 0; i < NET_IF_MAX_IPV4_MADDR; i++) {
+		if (iface->ipv4.mcast[i].is_used) {
+			continue;
+		}
+
+		iface->ipv4.mcast[i].is_used = true;
+		iface->ipv4.mcast[i].address.family = AF_INET;
+		iface->ipv4.mcast[i].address.in_addr.s4_addr32[0] =
+						addr->s4_addr32[0];
+
+		NET_DBG("[%d] interface %p address %s added", i, iface,
+			net_sprint_ipv4_addr(addr));
+
+		return &iface->ipv4.mcast[i];
+	}
+
+	return NULL;
+}
+
+bool net_if_ipv4_maddr_rm(struct net_if *iface, const struct in_addr *addr)
+{
+	int i;
+
+	for (i = 0; i < NET_IF_MAX_IPV4_MADDR; i++) {
+		if (!iface->ipv4.mcast[i].is_used) {
+			continue;
+		}
+
+		if (!net_ipv4_addr_cmp(
+			    &iface->ipv4.mcast[i].address.in_addr,
+			    addr)) {
+			continue;
+		}
+
+		iface->ipv4.mcast[i].is_used = false;
+
+		NET_DBG("[%d] interface %p address %s removed",
+			i, iface, net_sprint_ipv4_addr(addr));
+
+		return true;
+	}
+
+	return false;
+}
+
+struct net_if_mcast_addr *net_if_ipv4_maddr_lookup(const struct in_addr *maddr,
+						   struct net_if **ret)
+{
+	struct net_if *iface;
+
+	for (iface = __net_if_start; iface != __net_if_end; iface++) {
+		int i;
+
+		if (ret && *ret && iface != *ret) {
+			continue;
+		}
+
+		for (i = 0; i < NET_IF_MAX_IPV4_MADDR; i++) {
+			if (!iface->ipv4.mcast[i].is_used ||
+			    iface->ipv4.mcast[i].address.family != AF_INET) {
```

Comment:
How an ipv4 address's family could not be of AF_INET?
And actually, according to net_if_ipv4_maddr_add() above, it's always set to AF_INET. Plus, net_if_ipv4_maddr_rm() keep the family set, in the long run all maddr (used or not) will be of this family.

Looks like to me looping on iface->ipv4.mcast could be factorize for all the 3 functions, something like: loop_and_find_mcast(the mcast table, a bool used) returning the relevant mcast that matches used parameter, or NULL otherwise. It could be generalized about cmp also, but would require 2 more param: the maddr to compare with, and cmp boolean to tell which result is expected. if maddr is null, it will not run any comparison.

Shorter code, less imbrications of for/if/for etc...